### DISCUSSION TITLE
Is it impossible to run post-install scripts without self-hosting?

### DISCUSSION THREAD

**Discussion by serhalp** [renovatebot/renovate#11206] [Request Help] [answered]
Hi,

I've read every inch of the docs and countless Issues, PRs, and Discussions. I've tried a number of configurations to no avail.

Many of the hundreds of node repos in our org have hooks in `package.json#scripts` that are intended to be triggered automatically when `yarn install` is run. These may generate changes to additional files that must be committed along with the dependency bump. We would really like a single shared configuration in our org-wide preset that is referenced in all our repos that automatically handles this correctly. This feature was actually one of the reasons we chose to migrate from Dependabot to Renovate, and I'm disappointed to not be able to configure this so far. We use the Renovate GitHub app.

Here are all the things I've tried and why they failed.

### Solution 1: `allowScripts: true`

Because this is a "bot admin config" option (P.S.: yes, wow, please clarify this naming: https://github.com/renovatebot/renovate/issues/10887. It took me so long to figure out what this meant.), this option is quietly ignored when using the GitHub app.

### Solution 2: custom `postUpgradeTasks`

Say, if a repo has `package.json#scripts#prepare` of `foo && bar && baz`, then we could set in that repo's Renovate config:
```json
{
  "postUpgradeTasks": {
    "commands": ["foo && bar && baz"],
    "fileFilters": ["**/*"],
    "executionMode": "branch"
  }
}
```

This isn't particularly viable, because it requires me to configure this ad hoc in each repo, rather than in a shared org-wide configuration that "just works." It also requires duplicating the hook itself into the Renovate config.

This appears to not actually work. I _think_ this is because, despite `postUpgradeTasks` not being a "self-hosted configuration option", it only allows tasks that match the patterns in `allowedPostUpgradeCommands`, which _is_ a "self-hosted configuration option", and which presumably defaults to empty and therefore `postUpgradeTasks` isn't actually a usable option with the GitHub app(???).

### Solution 3: slightly less custom `postUpgradeTasks`

Say, if a repo has `package.json#scripts#prepare` of `foo && bar && baz`, then we could set in that repo's Renovate config:
```json
{
  "postUpgradeTasks": {
    "commands": ["yarn run prepare"],
    "fileFilters": ["**/*"],
    "executionMode": "branch"
  }
}
```

and if another repo has `package.json#scripts#postinstall` of `quux && foo`, then we could set in _that_ repo's Renovate config:
```json
{
  "postUpgradeTasks": {
    "commands": ["yarn run postinstall"],
    "fileFilters": ["**/*"],
    "executionMode": "branch"
  }
}
```

This is only slightly better than the previous solution, but it isn't particularly viable, since it requires me to configure this ad hoc in each repo, rather than in a shared org-wide configuration that "just works."

This option does not work in practice either, for the same reason as Solution 2.

### Solution 4: generic `postUpgradeTasks`

Specify this in our shared org-wide Renovate config:
```json
{
  "postUpgradeTasks": {
    "commands": ["yarn"],
    "fileFilters": ["**/*"],
    "executionMode": "branch"
  }
}
```

Since `yarn` (i.e. `yarn install`) is a no-op if it's already run and nothing needs to change, _but_ it always triggers post-install hooks regardless, this would be intended to just trigger the post-install hooks, regardless of which hook alias a particular repo is using (`prepare`, `postinstall`, legacy `prepublish` behavior that also triggers on post-install).

This sounds great to me – almost as simple as `allowScripts: true`, centralized in our org-wide config, and should just work.

Of course, this option does not work in practice either, for the same reason as Solutions 2 and 3.

## Bottom line

Is there no way whatsoever to accomplish this without self-hosting? Thanks in advance for any help you may provide! Renovate has been awesome so far, other than the frustration of working with the documentation. I'll open a couple other issues with some concrete feedback about that.

**Comment by viceice** [comment, marked as answer]
Yes, this only works for self-hosting. Thats why we have a separate docs page, where all admin options are written, which only works in bo config. And the bot config for the app is sealed by whitesource

**Reply by yetanotherion** [reply]
Thanks a lot @viceice!
For the ones that end up here too, FYI @viceice answered these questions at https://github.com/renovatebot/renovate/discussions/17906#discussioncomment-3699874 or https://github.com/renovatebot/renovate/discussions/16555#discussioncomment-3134107 too.

@viceice is there an offering provided by mend.io to access the botconfig in the github app after some subscription for ex? (There is no pricing mentioned at https://www.mend.io/free-developer-tools/contact-us/)

**Reply by rarkins** [reply]
We don't currently have any paid plans for Renovate. Arbitrary scripts are disabled for security reasons rather than commercial

**Reply by yetanotherion** [reply]
Thanks @rarkins! The same root cause is the reason why [customEnvVariables](https://github.com/renovatebot/renovate/discussions/20473) are not allowed either isn't it?

**Reply by rarkins** [reply]
Yes, although I think we could have an allow list/pattern for env variables. If we let every env variable through then there's some pretty wicked ways a user could interfere with a system

**Reply by yetanotherion** [reply]
Crystal clear, thanks again @rarkins!

**Comment by rarkins** [comment]
The docs are open source on this repo so you're welcome to simply propose changes directly with a PR if you prefer
